language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs

```diff
@@ -248,62 +242,75 @@ instead, but then we need to recover the original handy definitions and
 make the API and the code bloated (instances are more verbose with @DerivingVia@).
 -}
 
--- See Note [KnownType's defaults]
--- | Haskell types known to exist on the PLC side.
-class KnownType uni a where
+class KnownTypeAst uni a where
     -- | The type representing @a@ used on the PLC side.
     toTypeAst :: proxy a -> Type TyName uni ()
-    default toTypeAst :: KnownBuiltinType uni a => proxy a -> Type TyName uni ()
+    default toTypeAst :: uni `Includes` a => proxy a -> Type TyName uni ()
     toTypeAst _ = mkTyBuiltin @a ()
 
+-- | A constraint for \"@a@ is a 'KnownType' by means of being included in @uni@\".
+type KnownBuiltinType term a =
+    (HasConstant term, GShow (UniOf term), GEq (UniOf term), UniOf term `Includes` a)
+
+-- See Note [KnownType's defaults]
+-- | Haskell types known to exist on the PLC side.
+class KnownTypeAst (UniOf term) a => KnownType term a where
     -- | Convert a Haskell value to the corresponding PLC term.
     -- The inverse of 'readKnown'.
-    makeKnown :: a -> Term TyName Name uni ()
-    default makeKnown :: KnownBuiltinType uni a => a -> Term TyName Name uni ()
+    makeKnown :: a -> EvaluationResult term
+    default makeKnown :: KnownBuiltinType term a => a -> EvaluationResult term
```

Comment:
Now we don't turn `EvaluationFailure` into `Error` by `makeKnown`. The reason for that is that Untyped Plutus Core will have its own `Error`, so we'd have to either handle this with a class or just return `EvaluationResult` directly. I tried the latter option and it worked well, so I think it makes sense to stick to it.